converter/src/main/java/org/dswarm/converter/flow/TransformationFlow.java

```diff
@@ -221,129 +218,92 @@ public String apply(
-			if (gdmModel.getRecordClassURI() == null) {
+						final String recordURI = gdmModel.getRecordURIs().iterator().next();
 
-				final String recordURI = gdmModel.getRecordURIs().iterator().next();
+						final Resource recordResource = state.model.getResource(recordURI);
 
-				final Resource recordResource = model.getResource(recordURI);
+						if (recordResource != null) {
 
-				if (recordResource != null) {
+							// TODO check this: subject OK?
+							recordResource.addStatement(new ResourceNode(recordResource.getUri()), new Predicate(GDMUtil.RDF_type),
+									new ResourceNode(defaultRecordClassURI));
+						}
 
-					// TODO check this: subject OK?
-					recordResource.addStatement(new ResourceNode(recordResource.getUri()), new Predicate(GDMUtil.RDF_type),
-							new ResourceNode(defaultRecordClassURI));
-				}
+						// re-write GDM model
+						finalGDMModel = new GDMModel(gdmModel.getModel(), recordURI, defaultRecordClassURI);
+					} else {
 
-				// re-write GDM model
-				finalGDMModel = new GDMModel(gdmModel.getModel(), recordURI, defaultRecordClassURI);
-			} else {
+						finalGDMModel = gdmModel;
+					}
 
-				finalGDMModel = gdmModel;
-			}
+					return state
+							.setRecordClassUriIfUnset(finalGDMModel.getRecordClassURI())
+							.addModel(finalGDMModel)
+							.addRecordUris(recordURIsFromGDMModel);
+				})
+				.map(state -> {
+					if (state.recordClassUri == null && optionalDataModelSchemaRecordClassURI.isPresent()) {
+						state.setRecordClassUriIfUnset(optionalDataModelSchemaRecordClassURI.get());
+					}
+					return state;
+				});
 
-			if (recordClassUri == null) {
-
-				recordClassUri = finalGDMModel.getRecordClassURI();
-			}
-
-			finalGDMModels.add(finalGDMModel);
-
-			recordURIs.addAll(recordURIsFromGDMModel);
-		}
-
-		if (recordClassUri == null && optionalDataModelSchemaRecordClassURI.isPresent()) {
-
-			// set data model schema record class URI
-			recordClassUri = optionalDataModelSchemaRecordClassURI.get();
-		}
-
-		// note: we may don't really need the record class uri here (I guess), because we can provide the record identifiers
-		// separately
-		final GDMModel gdmModel = new GDMModel(model, null, recordClassUri);
-		gdmModel.setRecordURIs(recordURIs);
+		final Observable<GDMTransformationState> resultObservable;
 
 		if (writeResultToDatahub) {
 
-			if (outputDataModel.isPresent() && outputDataModel.get().getUuid() != null) {
+			if (hasDefined(outputDataModel, DMPObject::getUuid)) {
 
 				// write result to graph db
 				final InternalModelService internalModelService = internalModelServiceFactoryProvider.get().getInternalGDMGraphService();
 
-				try {
+				resultObservable = stateObservable.map(state -> {
 
-					internalModelService.createObject(outputDataModel.get().getUuid(), gdmModel);
-				} catch (final DMPPersistenceException e1) {
+					try {
+						internalModelService.updateObject(outputDataModel.get().getUuid(), state.finalModel(), UpdateFormat.DELTA, true);
```

Comment:
we should probably make the update a bit more stream friendly (in the future). right now, we write the whole model at once.